### PR TITLE
ZipExtractionInstaller should perform similar checks to DownloadFromUrlInstaller to prevent multiple unnecessary installs

### PR DESCRIPTION
**Pull request by itsjohno**
Minor quality of life improvement for ZipExtractionInstaller _(borrowing logic from DownloadFromUrlInstaller.java)_. Issue found during installation of tools from local FTP machine, and Jenkins tries to install the tool multiple times.

See related Google Groups post: https://groups.google.com/forum/#!topic/jenkinsci-users/FBR7v9raztQ

_**This is my first pull request, so please let me know if I've done anything wrong!**_

### Proposed changelog entries

* Track installation of tools using ZipExtractionInstaller using same method as DownloadFromUrlInstaller to prevent multiple installations

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Comment by itsjohno** [comment]
Agh! Test failures. I shall look into this.

**Comment by daniel-beck** [comment]
The problem here is, it's reasonable to expect that the same binary for a specific version distributed by a tool project doesn't get replaced. The same cannot be said for zip files distributed internally. There may just be one `tools.zip` that gets updated as needed.

**Comment by itsjohno** [comment]
@daniel-beck I think that's a fair point, I was trusting that it would be a different URL if someone was changing the contents. I like the idea of using a checksum but I can't see a way that you could generate the checksum without fetching the entire tool zip file and then generating.

The way we were using this internally was to place tools on an internal FTP which meant the Jenkins could fetch Maven and JDK, without having to go externally. In essence, just providing an explicit URL to fetch a specific version of a tool (and as you said, this should not change and thus not be replaced).

File: core/src/main/java/hudson/tools/ZipExtractionInstaller.java [hunk 1]
```diff
@@ -90,17 +95,25 @@ public FilePath performInstallation(ToolInstallation tool, Node node, TaskListen
...
     @Extension @Symbol("zip")
     public static class DescriptorImpl extends ToolInstallerDescriptor<ZipExtractionInstaller> {
 
         public String getDisplayName() {
             return Messages.ZipExtractionInstaller_DescriptorImpl_displayName();
         }
 
-        @RequirePOST
         public FormValidation doCheckUrl(@QueryParameter String value) {
-            Jenkins.getInstance().checkPermission(Jenkins.ADMINISTER);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tools/ZipExtractionInstaller.java]: this is a bad idea (CSRF risk, DDoS)

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
This change blocks tool updates as discussed above. 
I do not think there is a good way to change the current installer reliably. 

Instead of the change, I would rather recommend creating a new Tool installer in https://plugins.jenkins.io/extra-tool-installers . I will be happy to merge and release it if it is properly documented

File: core/src/main/java/hudson/tools/ZipExtractionInstaller.java (hunk 1 above)
Comment (itsjohno) [reply core/src/main/java/hudson/tools/ZipExtractionInstaller.java]: Not sure how this line ended up getting pulled out (explains the failing UT), guess when I sync'd upstream it didn't get pulled in right.

**Comment by itsjohno** [comment]
Happy to try doing this as a tool installer in the plugin, I'll close this PR and give that a shot. Thanks @oleg-nenashev :)

**Comment by itsjohno** [comment]
As an aside though, I agree about the tool upgrade path but then it doesn't seem right that it would re-run the installer on every single job run? Does anyone have any ideas as to how that could be tackled? I like the idea of checksums but not sure how that could be implemented? Perhaps verifying every hour or so? (or making it configurable?)

**Comment by oleg-nenashev** [comment]
In a new installer you can use the path-based approach. It's fine if it's documented correctly.
Ideally we could cache ZIPs on master, calculate checksums there, and then check for updates on the remote side periodically. Agents would be pulling the master's version then.

**Comment by daniel-beck** [comment]
Checksums, or if we're feeling slightly adventurous, Last-Modified headers (or both) could be archived similar to the current cache (of URL IIRC?) and used for subsequent uses. Either should cut the number of unnecessary installs by a lot, if perhaps not downloads (in the case of checksums).

So there's a way forward here, but it's a bit more work.

**Comment by itsjohno** [comment]
@oleg-nenashev I like the idea of storing a cached version on the master which is then pushed if there's changes (since it means we can incorporate @daniel-beck's checksums/last-modified headers approach). I'll investigate a bit further and open a new PR when ready :)

**Pull request closed without merging**